### ISSUE TITLE
EventPipe crash during DestroyThread 

### ISSUE THREAD

**Issue by chunseoklee** [dotnet/coreclr#15436]
Current DestroyThread function is like below. 
https://github.com/dotnet/coreclr/blob/master/src/vm/threads.cpp#L907
```
void DestroyThread(Thread *th)
...
    if (g_fEEShutDown == 0) 
    {
        th->SetThreadState(Thread::TS_ReportDead);
        th->OnThreadTerminate(FALSE);
    }

#ifdef FEATURE_PERFTRACING
    // Before the thread dies, mark its buffers as no longer owned
    // so that they can be cleaned up after the thread dies.
    EventPipeBufferList *pBufferList = th->GetEventPipeBufferList();
    if(pBufferList != NULL)
    {
        pBufferList->SetOwnedByThread(false);
    }
#endif // FEATURE_PERFTRACING
}
```

On tizen arm machine, seg fault occurs at https://github.com/dotnet/coreclr/blob/master/src/vm/threads.cpp#L1019.
```
EventPipeBufferList *pBufferList = th->GetEventPipeBufferList();
```

Is this normal flow during DestroyThread ? I am not familair with semantic of destroying thread.

**Comment by chunseoklee** [comment]
@janvorli @jkotas PTAL

**Comment by jkotas** [comment]
cc @brianrob

**Comment by brianrob** [comment]
@chunseoklee, just to confirm, this is the line where the AV occurs, right?

```
  EventPipeBufferList *pBufferList = m_pEventPipeBufferList.Load();
```

m_pEventPipeBufferList is of type Volatile<EventPipeBufferList*>.  It looks like VolatileLoad, which is what gets called by Load has platform specific implementations.

Are you seeing this on ARM32 or ARM64?

**Comment by chunseoklee** [comment]
Oh. As you mentioned, av occurs inside VolatileLoad. 
And ARM32. 
#15437 seems to fix this issue, now. 
Thank you

**Comment by brianrob** [comment]
Thanks for the info.

And @kbaladurin thanks for the fix.

**Comment by sherlock1982** [comment]
Any news if it can get to release/2.0.0 ?
I've seen this today on Stretch x64 NETCore 2.0.3:

```
#0  VolatileLoad<EventPipeBufferList*> (pt=<optimized out>) at /root/coreclr/src/inc/volatile.h:123
#1  Volatile<EventPipeBufferList*>::Load (this=<optimized out>) at /root/coreclr/src/inc/volatile.h:248
#2  Volatile<EventPipeBufferList*>::operator EventPipeBufferList* (this=<optimized out>) at /root/coreclr/src/inc/volatile.h:302
#3  Thread::GetEventPipeBufferList (this=0x2bc2da0) at /root/coreclr/src/vm/threads.h:5357
#4  DestroyThread (th=0x2bc2da0) at /root/coreclr/src/vm/threads.cpp:999
#5  0x00007faa9015055a in BackgroundThreadStub (arg=<optimized out>) at /root/coreclr/src/vm/gcenv.ee.cpp:808
#6  0x00007faa90496952 in CorUnix::CPalThread::ThreadEntry (pvParam=0x7fa9d80086c0) at /root/coreclr/src/pal/src/thread/thread.cpp:1749
#7  0x00007faa91885494 in start_thread (arg=0x7fa9c9573700) at pthread_create.c:333
#8  0x00007faa90d2aaff in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97

```

Looks like it's the same issue

**Comment by janvorli** [comment]
@brianrob - do you plan to port the fix?

**Comment by brianrob** [comment]
Thanks for reporting @sherlock1982.

@janvorli, yup, just sent another mail with more information requested by shiproom.

**Comment by brianrob** [comment]
The fix for this has been approved for a 2.0 servicing release.  I will work on getting it merged.

**Comment by brianrob** [comment]
This is already fixed in master and has been approved for porting to release/2.0.  Closing this issue.
